### ISSUE TITLE
hashsum.rs uses read_to_end incorrectly, allocates more memory than it intends

### ISSUE THREAD

**Issue by oconnor663** [uutils/coreutils#2295]
In the `digest_reader` function in `hashsum.rs` there are these lines:

```rust
// Digest file, do not hold too much in memory at any given moment
let mut buffer = Vec::with_capacity(524_288);
```

However, right below that, we call `reader.read_to_end(&mut buffer)`. I think the author of this code might've thought that "read to end" means "read up to the capacity of the current vector," but that's not what it means. It means "read to the _end of the file_, allocating more capacity if needed." That line of code reads the entire input into memory the first time it runs.

We can confirm this with a couple of `dbg!` calls:

```diff
diff --git a/src/uu/hashsum/src/hashsum.rs b/src/uu/hashsum/src/hashsum.rs
index 2e31ddd2..54219b0a 100644
--- a/src/uu/hashsum/src/hashsum.rs
+++ b/src/uu/hashsum/src/hashsum.rs
@@ -559,6 +559,7 @@ fn digest_reader<'a, T: Read>(
     // Digest file, do not hold too much in memory at any given moment
     let windows = cfg!(windows);
     let mut buffer = Vec::with_capacity(524_288);
+    dbg!(buffer.capacity());
     let mut vec = Vec::with_capacity(524_288);
     let mut looking_for_newline = false;
     loop {
@@ -593,6 +594,7 @@ fn digest_reader<'a, T: Read>(
             Err(e) => return Err(e),
         }
     }
+    dbg!(buffer.capacity());
     if windows && looking_for_newline {
         vec.push(b'\r');
         digest.input(&vec);
```

When I hash a 1 GB file, here's what I see:

```
> ~/uutils-coreutils/target/release/coreutils hashsum --b2sum f
[src/uu/hashsum/src/hashsum.rs:562] buffer.capacity() = 524288
[src/uu/hashsum/src/hashsum.rs:597] buffer.capacity() = 1073741824
3267363e20da32499ccbaa49da9a70d62ec19b4b0d863edc3d12f9d8eb9eaf9747a81e58d8628858a3966bf4be2c6531f351037bbd46f0fddcb42c629d5ac3dd  f
```

So indeed, this code has allocated 1 GB of RAM.

There are various ways to fix this, but overall the whole input loop here feels odd to me. It's trying to accomodate text mode reads on Windows, but I'm not sure it's possible to do this unambiguously. (Apologies, I don't have a Windows machine with me at the moment to test it.)

**Comment by miDeb** [comment]
From briefly looking at this, the fix for this would be to replace the `read_to_end` call with a simple `read` and pre-fill the vec with zeroes (replace `Vec::with_capacity(capacity)` with `vec![0; capacity]`. I think there are some more problems, however:
- We don't need an additional `BufReader`. `digest_reader` already creates its own buffer, so right now there are two buffers.
- The buffer size seems too big. We could start with a smaller buffer and maybe grow it later. That's something to be benchmarked, though. I'd also check if we could benefit from splitting the task of reading from the file and hashing between two threads.
- If we try to parse in windows text mode (whatever that is - there are no tests for it and I have no idea why we want to modify the input file before hashing), there is a second vector used. That allocation is not used in other cases, we should make sure it's only allocated when actually used.

**Comment by jfinkels** [comment]
I believe this is fixed by pull request #2663.
